### PR TITLE
Create oc_presets

### PR DESCRIPTION
**Pull request by RomkaSwallowtail**
A new plugin allows you to save the current size and position of the collar in presets.
Then, if necessary, restore the desired size / position from the saved one.
You can save multiple sets of collar sizes / positions for different costumes and / or avatars.

You can send the “preset 'set name'” command to the collar via the chat channel or g_iHUDChan channel:

string preset = "werwolf";
key g_kWearer = llGetOwner();
integer g_iHUDChan = -llAbs((integer)("0x"+llGetSubString((string)g_kWearer,-7,-1)));
llRegionSayTo(g_kWearer, g_iHUDChan, (string)g_kWearer + ":restore "+preset);

### CODE REVIEW

**Comment by zontreck** [comment]
We'll have to discuss the addition of this script. We presently have in the works a themes replacement and I believe it does or will support this functionality among other things. At the moment I'm not going to merge. We need to establish whether this script will be needed with the themes replacement we're going to be pushing out.

File: src/spares/oc_presets [hunk 1]
```diff
@@ -0,0 +1,220 @@
+ /*
+ Copyright (c) 2019 Romka Swallowtail
+
+ Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (zontreck) [inline src/spares/oc_presets]: We prefer GPLv2. If possible can you change this?

**Comment by zontreck** [comment]
Changes Requested

Please see the review for details.

File: src/spares/oc_presets (hunk 1 above)
Comment (RomkaSwallowtail) [reply src/spares/oc_presets]: yes

Comment (RomkaSwallowtail) [reply src/spares/oc_presets]: I did

**Comment by zontreck** [comment]
Testing script. If more changes are needed i will post here in the next few minutes.

File: src/spares/oc_presets [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    } else if (sType == DEL) sPrompt = "\nSelect preset or create new to delete size/position/rotation";
+    if (sType == NEW) {
+        sPrompt = "\nType new preset name";
+        lMyButtons = [];
+        lUtils = [];
+    }
+    Dialog(kAv, sPrompt, lMyButtons, lUtils, 0, iAuth, sType);
+}
+
+Restore(string sName, key kAv) {
```
Comment (zontreck) [inline src/spares/oc_presets]: This appears to do nothing when I attempted to restore after having resized the collar and moved several of the prims around. Various other configurations did not result in restore working.

**Comment by zontreck** [comment]
Problematic segments
=======

* Restore
==
<details>
<summary>Snippet from Restore(string,key)</summary>

```

        vector vDestSize = (vector)llList2String(lSettings,0);
        float fScale = vDestSize.x / vCurSize.x;
        llScaleByFactor(fScale);
        llSetPos((vector)llList2String(lSettings,1));
        llSetLocalRot((rotation)llList2String(lSettings,2));

```
</details>

The above code will only restore the root prim. As many collars are >1 prim, this would need to be adjusted. Take into account also the oc_settings memory limits, too many presets and oc_settings will stackheap.

<details>
<summary>Snippet from Save(string,key)</summary>

```

Save(string sName, key kAv) {
    string sSettings = (string)llGetScale()+"/"+(string)llGetLocalPos()+"/"+(string)llGetLocalRot();
```
</details>

The above will only save for the most part, for the root prim, as each prim has its own scale, position and rotation.

**Comment by zontreck** [comment]
Incase you have trouble reproducing the Restore function not working-

I created 3 presets. 

* Default
* Changed
* Changed2

All 3 had different properties; And I did attempt restoring each preset.

File: src/spares/oc_presets (hunk 2 above)
Comment (RomkaSwallowtail) [reply src/spares/oc_presets]: did You do SAVE before?
Collar menu -> Apps -> SizePresets -> SAVE ,  type preset name in textbox and press Submit button to save current Collar size/position/rotation
after that, the saved preset will appear in the RESTORE menu

You can change the size and position of the collar through the Settings menu, and then save it as a new preset

**Comment by RomkaSwallowtail** [comment]
llScaleByFactor(fScale)  Attempts to resize the entire object by scaling_factor, maintaining the size-position ratios of the prims.  **All object, not only root prim!**

No need save all prims size/position/rotation! root prim is used as a reference!
it works in my collar! works well, tested on cats! :)

**Comment by zontreck** [comment]
Okay, i'd still need to see a working script though. I sent you a IM inworld, but you were offline. The save and delete functions do work, however for some reason during testing Restore did not.

File: src/spares/oc_presets (hunk 2 above)
Comment (zontreck) [reply src/spares/oc_presets]: See my comment about what steps I took. I saved three different presets

**Comment by zontreck** [comment]
Thank you for your contribution. I overlooked the attached requirement. Will discuss with other staff about whether we need this to be more verbose to avoid it being mistaken for not working.

**Pull request merged**